### DISCUSSION TITLE
asyncpg's CannotConnectNowError not wrapped/caught on connect

### DISCUSSION THREAD

**Discussion by timwoocker** [sqlalchemy/sqlalchemy#11954] [Usage Questions] [answered]
When trying to run `engine.connect()` with asyncpg set as the driver, establishing the connection might fail due to the `CannotConnectNowError` exception. This exception indicates that postgres is currently booting up and will most likely be available a few milliseconds later.

When using SQLAlchemy, I would expect one of these two:
1. the exception is wrapped in `DisconnectionError ` making it easy for me to catch and handle the exception myself
2. the exception is caught by SQLAlchemy and it automatically tries to reconnect (apparently that's the default behaviour for `DisconnectionError` anyways so the above is probably what the driver should do)

However right now the exception is not wrapped so I need to catch a specific asyncpg exception which means my code now relies on asyncpg - I would not be able to switch out the driver later because I need to catch a specific asyncpg exception.

Is this something that should be fixed or is this expected behaviour for some reason?

**Comment by CaselIT** [comment]
Hi,

It seems like a bug. ~~It would be great if you cold provide a PR to fix~~  edit by mike:  see #11956, this is a bigger issue

**Reply by timwoocker** [reply]
Thank you for the reply.

I found the part in the code where the exception would need to be caught but the code on the main branch differs from the newest 2.0.35 release even though the lines have been edited 10 months ago: https://github.com/sqlalchemy/sqlalchemy/blame/rel_2_0_35/lib/sqlalchemy/dialects/postgresql/asyncpg.py#L946 and https://github.com/sqlalchemy/sqlalchemy/blame/main/lib/sqlalchemy/dialects/postgresql/asyncpg.py#L961

I don't seem to understand which version of the code I should touch when trying to solve this issue. I mean I would create a new branch off the main branch but that's not the code in the release even though it's older.

**Reply by CaselIT** [reply]
main is the one to target. Then a maintainer will backport to 2_0 if appropriate  (as it would be in this case)

**Comment by zzzeek** [comment]
hi -

`CannotConnectNowError` extends from `PostgresError`:

```
>>> asyncpg.exceptions.CannotConnectNowError.__mro__
(<class 'asyncpg.exceptions.CannotConnectNowError'>, <class 'asyncpg.exceptions.OperatorInterventionError'>, <class 'asyncpg.exceptions._base.PostgresError'>, <class 'asyncpg.exceptions._base.PostgresMessage'>, <class 'Exception'>, <class 'BaseException'>, <class 'object'>)

```

we should be wrapping `CannotConnectNowError` as a result since we walk through `__mro__`.

however, the `is_disconnect` decision is separate so that would be parsed here:

https://github.com/sqlalchemy/sqlalchemy/tree/main/lib/sqlalchemy/dialects/postgresql/asyncpg.py#L1165-L1171

**Reply by timwoocker** [reply]
I don't quite understand what you mean. Exceptions raised in connect() are not handled anywhere so is_disconnect would not know about the exception.

I'm sorry if I miss something - this is my first time actually looking into the source of SQLAlchemy

**Reply by zzzeek** [reply, marked as answer]
oh, in connect.   the async DBAPIs dont wrap the actual "connect()" right now, and that's a bug.  This is a major oversight in all the async DBAPIs and we will have to do this for 2.1, this is #11956. 


however, that will not get you the behavior you are requesting:

> 1. the exception is wrapped in `DisconnectionError ` making it easy for me to catch and handle the exception myself

this is not a disconnect, this a connection failed.  These don't get any kind of "disconnection error" scenario nor is the "connection_invalidated" flag set because there is no connection:

```py
>>> from sqlalchemy import create_engine
>>> e = create_engine("mysql://scott:tiger12@localhost/test")   # note this is the wrong password
>>> try:
...     conn = e.connect()
... except Exception as e:
...     err = e
... 
>>> err
OperationalError('(MySQLdb.OperationalError) (1045, "Access denied for user \'scott\'@\'localhost\' (using password: YES)")')
>>> err.connection_invalidated
False

```

So above, you'd get something like an `OperationalError` when we implement #11956 but you'd still need to parse its contents if you want to distinguish between specific kinds of failure modes on connect (which you don't have to, for your use case, you just should keep trying to connect).

in the interim you just have to add the asyncpg exception to your catch, as well as that of any other async DBAPIs you're using, since #11956 is likely too big for 2.0.x.

> 2. the exception is caught by SQLAlchemy and it automatically tries to reconnect (apparently that's the default behaviour for DisconnectionError anyways so the above is probably what the driver should do)

SQLAlchemy does not do this when it first connects.   There is only a "three attempts" logic for existing connections that were already made and are being re-checked out from the pool, which takes place either when using pre-ping logic, or when the [checkout](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.checkout) event is used and raises `DisconnectionError`.

**Reply by timwoocker** [reply]
Thank you for looking into this!
I'm already catching the asyncpg exception as a workaround. It's great to know this will be fixed in a future release!
I would love to help with this but seems like a relatively large refactore and I don't know the source good enough I'm afraid.
